### PR TITLE
feature/seattle-sara-nelson

### LINKED ISSUE #57: Add Sara Nelson to Seattle Council Static
We got a new council member in seattle: http://www.seattle.gov/council/meet-the-council/sara-nelson

Can we add her info to the static file before I ship a new version of this lib.

### PR DESCRIPTION
**Pull request by dphoria**
### Link to Relevant Issue

This pull request resolves #57 

### Description of Changes

_Include a description of the proposed changes._

Added new Seattle City Council member Sara nelson to `seattle-static.json`.

### CODE REVIEW

**Comment by dphoria** [comment]
/test-scraper get_seattle_events --from_dt=2022-01-04T00:00:00 --to_dt=2022-01-05T00:00:00

**Comment by dphoria** [comment]
Sara Nelson now included:

```Json
{
    "name": "Sara Nelson",
    "is_active": true,
    "router_string": null,
    "email": "Sara.Nelson@seattle.gov",
    "phone": null,
    "website": "http://www.seattle.gov/council/meet-the-council/sara-nelson",
    "picture_uri": "http://www.seattle.gov//assets/images/Council/Members/Nelson/Councilmember-Sara-Nelson_homepage-banner.jpg",
    "seat": {
        "name": "Position 9",
        "electoral_area": "Citywide",
        "electoral_type": null,
        "image_uri": "https://upload.wikimedia.org/wikipedia/commons/3/36/SeattleI5Skyline.jpg",
        "external_source_id": null,
        "roles": [
            {
                "title": "Councilmember",
                "body": {
                    "name": "City Council",
                    "is_active": true,
                    "start_datetime": null,
                    "description": null,
                    "end_datetime": null,
                    "external_source_id": "138"
                },
                "start_datetime": 1641024000.0,
                "end_datetime": 1735632000.0,
                "external_source_id": "1195"
            },
            {
                "title": "Member",
                "body": {
                    "name": "Council Briefing",
                    "is_active": true,
                    "start_datetime": null,
                    "description": null,
                    "end_datetime": null,
                    "external_source_id": "211"
                },
                "start_datetime": 1641024000.0,
                "end_datetime": 1735632000.0,
                "external_source_id": "1198"
            },
            {
                "title": "Vice Chair",
                "body": {
                    "name": "Sustainability and Renters' Rights Committee",
                    "is_active": true,
                    "start_datetime": null,
                    "description": null,
                    "end_datetime": null,
                    "external_source_id": "252"
                },
                "start_datetime": 1641283200.0,
                "end_datetime": 1704009600.0,
                "external_source_id": "1220"
            },
            {
                "title": "Member",
                "body": {
                    "name": "Public Safety and Human Services Committee",
                    "is_active": true,
                    "start_datetime": null,
                    "description": null,
                    "end_datetime": null,
                    "external_source_id": "251"
                },
                "start_datetime": 1641283200.0,
                "end_datetime": 1704009600.0,
                "external_source_id": "1229"
            },
            {
                "title": "Chair",
                "body": {
                    "name": "Economic Development, Technology, and City Light Committee",
                    "is_active": true,
                    "start_datetime": null,
                    "description": null,
                    "end_datetime": null,
                    "external_source_id": "259"
                },
                "start_datetime": 1641369600.0,
                "end_datetime": 1704009600.0,
                "external_source_id": "1236"
            },
            {
                "title": "Member",
                "body": {
                    "name": "Neighborhoods, Education, Civil Rights, and Culture Committee",
                    "is_active": true,
                    "start_datetime": null,
                    "description": null,
                    "end_datetime": null,
                    "external_source_id": "261"
                },
                "start_datetime": 1641369600.0,
                "end_datetime": 1704009600.0,
                "external_source_id": "1249"
            },
            {
                "title": "Member",
                "body": {
                    "name": "Land Use Committee",
                    "is_active": true,
                    "start_datetime": null,
                    "description": null,
                    "end_datetime": null,
                    "external_source_id": "262"
                },
                "start_datetime": 1641283200.0,
                "end_datetime": 1704009600.0,
                "external_source_id": "1255"
            }
        ]
    },
    "external_source_id": "706"
}
```

**Comment by dphoria** [comment]
CC @isaacna @tohuynh in case you're interested in our new member in Seattle ingestion data! :sweat_smile:

**Comment by dphoria** [comment]
FYI for the "future" reader, I did nothing manual. Ran `SeattleScraper.get_static_person_info()` which pulled information for the new member out of the box. Copied the json blob for this person out of the returned data set into `seattle-static.json`, then simply reused existing `Seat.image_uri` for that district.

TLDR, methods like `SeattleScraper.get_static_person_info()` can be used to get new council member(s) information.

**Review by isaacna — COMMENTED** [review]
The new data looks good to me! But don't we have to change Gonzalez's `is_active` to `False` since Sara Nelson replaced her?

Or I guess I should say, should the scraper/something else be detecting when a position is changed, and set the old holder of that position to inactive?

**Comment by dphoria** [comment]
You make a good point. The scraper is simply using Legstar's `PersonActiveFlag`. So I kind of like your suggestion of, when multiple people are found to hold the same position, make the "previous" person inactive. Does that align with your suggestion @isaacna ?

**Comment by isaacna** [comment]
> when multiple people are found to hold the same position, make the "previous" person inactive. Does that align with your suggestion @isaacna ?

Yep exactly. I noticed in #57 @JacksonMaxfield mentioned to not replace the data for Gonzalez, but I'm not sure if this meant fully replaced and whether toggling `is_active` is ok.

**Comment by tohuynh** [comment]
Just a question for @JacksonMaxfield, not related to this PR, but I see the roles generated here.

Is the `end_datetime` on these `Role`s supposed to be null?

>This is the correct filter I believe. `end_datetime` == `null` means the person is active in council.
The behavior Isaac and I discussed when we made that collection was that `null` meant they are in office. Every four years when an election happens, the maintainer of the instance can update either the scraper with new info: https://github.com/CouncilDataProject/cdp-scrapers/blob/main/cdp_scrapers/instances/seattle-static.json#L89
Or just update the database itself.

_Originally posted by @JacksonMaxfield in https://github.com/CouncilDataProject/cdp-frontend/pull/122#discussion_r760643039_

FYI, current roles for both seattle-staging and king-county all have non-null `end_datetime`. (that why there are no currently elected persons on the `/people` page because cdp-frontend is using the null check).

**Comment by dphoria** [comment]
Just to chime in, we get `Role.end_datetime` dynamically from Legistar `OfficeRecordEndDate`.

**Comment by tohuynh** [comment]
Looks like we ended up using the ingestion role's `end_datetime`, regardless of whether the `end_datime` is > `now()` or not.
 https://github.com/CouncilDataProject/cdp-backend/blob/main/cdp_backend/database/functions.py#L418

**Comment by JacksonMaxfield** [comment]
Sorry for being away for a bit! Yes @isaacna @dphoria Gonzalez should be kept but should be marked as inactive. Her roles should have an end datetime of 2021-12-31 if possible.

@tohuynh the end datetime of null means it's an active role.

**Review by JacksonMaxfield — APPROVED** [review]
Sara Nelson's stuff looks good but can we set Gonzalez to inactive?

Interestingly, to update her roles I think we need to have like a different metadata pipeline or something because as a reminder this info only gets injected when the person is present / voted in an event.

We may need to write a quick "just update the metadata" function or something.

**Comment by dphoria** [comment]
I'll include changes for Gonzalez in this PR; involves the same `seattle-static.json` file.

Nelson is now in votes for events like the one tested above our from 01/04. I see what you mean; no way to just pull and update person information. 

I can add a `person`-only-getter to the scrapers. Would this work for this new pipeline?

**Comment by dphoria** [comment]
Wonder if there is other information besides `Person` that would be nice to get outside of `get_events()`.

**Comment by JacksonMaxfield** [comment]
> I can add a `person`-only-getter to the scrapers. Would this work for this new pipeline?

> Wonder if there is other information besides Person that would be nice to get outside of get_events().

Definitely matter info. Like if a bill has been signed by mayor / if a bill has been introduced to council but hasn't yet been to an event.

**Comment by dphoria** [comment]
This conversation is starting to deviate from this PR :sweat_smile: so I created an issue. https://github.com/CouncilDataProject/cdp-scrapers/issues/59

**Comment by tohuynh** [comment]
> @tohuynh the end datetime of null means it's an active role.

Then some of the active roles' `end_datetime` need to be changed to null in the DB.

And change this [line of code](https://github.com/CouncilDataProject/cdp-backend/blob/main/cdp_backend/database/functions.py#L418) to use null when the ingestion role's `end_datetime` is greater now() ?

**Comment by JacksonMaxfield** [comment]
> Then some of the active roles' `end_datetime` need to be changed to null in the DB.
> 
> And change this [line of code](https://github.com/CouncilDataProject/cdp-backend/blob/main/cdp_backend/database/functions.py#L418) to use null when the ingestion role's `end_datetime` is greater than now() ?

Sorry, clarifying: end datetimes that are > now() or are null should be considered active roles. The reasoning behind this was simply "make it easy on the instance maintainer to decide how to handle role info..."

What roles need to be changed?

**Comment by tohuynh** [comment]
> end datetimes that are > now() or are null should be considered active roles. 

Is there a way to get active roles on the frontend, given this condition? If not I suppose we can get all roles and then filter on the frontend.

**Comment by dphoria** [comment]
Would [this](https://drive.google.com/file/d/1PR7cZni7zZTypRGj43qQsDq1iLlwgIHl/view?usp=sharing) might help with the conversation? This is what Gonzalez's roles are in Legistar these days. i.e. `Role.end_datetime` would reflect these values if an event that includes her as sponsor or voter is pulled. (Or by using the upcoming new API).

The caveat is that the municipality HAS updated the info on their Legistar DB. I do have an idea that would allow any instance to specify "override" values for all roles of a given person. e.g. Use 12/31/2021 for end_datetime for all roles for Gonzalez. Not sure yet if such functionality would indeed be necessary.

**Comment by JacksonMaxfield** [comment]
Those roles from legistar look great. All of the end datetimes I see are 2021-12-31 which is perfect. Seems like the new API + pipeline would clean that up.

**Comment by JacksonMaxfield** [comment]
> Is there a way to get active roles on the frontend, given this condition? If not I suppose we can get all roles and then filter on the frontend.

When do we need roles? Only on the individual person page right? In which case that should be a cheap call to make for a single person.

**Comment by tohuynh** [comment]
Yes, end_datetime that are currently in the DB is perfect. What I'm concerned with is having one way to query for active roles. For example, a query with end_datime > now() would not include null end_datetimes and a query with end_datime == null would not include the active roles with non-null end_date.time.

> When do we need roles? Only on the individual person page right? In which case that should be a cheap call to make for a single person.

Also the `/people` page

**Comment by tohuynh** [comment]
Or we can make two separate queries? I think that could be fine.

**Comment by JacksonMaxfield** [comment]
> Or we can make two separate queries? I think that could be fine.

I think for now that is the easiest thing. From an API perspective / data cleaning perspective I see what you mean. It's annoying. Idk if I really want to fix it on the database model side though... I truly see ups and downs.

**Review by JacksonMaxfield — APPROVED** [review]
Thanks for following and encouraging the long conversation AND making the changes!

We should talk about the new pipeline in meeting tomorrow!

Feel free to merge whenever.

**Comment by dphoria** [comment]
Roles for Gonzalez should now reflect `end_datetime` provided by Seattle Legistar. Will run test to make sure. e.g. Something like

```
Role(
    title='Member',
    body=Body(name='Select Committee on Homelessness Strategies and Investments', ...),
    start_datetime=datetime.datetime(2020, 1, 7, 0, 0, tzinfo=<DstTzInfo 'America/Los_Angeles' PST-1 day, 16:00:00 STD>),
    end_datetime=datetime.datetime(2021, 12, 31, 0, 0, tzinfo=<DstTzInfo 'America/Los_Angeles' PST-1 day, 16:00:00 STD>),
    external_source_id='1162'
)
```

**Comment by dphoria** [comment]
/test-scraper get_seattle_events --from_dt=2021-12-15T00:00:00 --to_dt=2021-12-16T00:00:00

**Comment by dphoria** [comment]
> `/test-scraper` is finished. Please go to the [Actions page](https://github.com/CouncilDataProject/cdp-scrapers/actions) then click on the latest workflow titled "feature/seattle-sara-nelson". The results are saved in `scraper-test-results` under Artifacts on the workflow summary page.

`Role.end_datetime`s for Gonzalez confirmed.
